src/components/App/App.js

```diff
@@ -0,0 +1,362 @@
+  const [userData, setUserData] = React.useState({username: "Foma Kiniaev"})
+
+  //search
+
+  const [isLoading, setIsLoading] = React.useState(false);
+  const [foundArticles, setFoundArticles] = React.useState([]);
+  const articles = React.useRef([]);
+
+  //popups
+
+  const [isLoginPopupOpen, setIsLoginPopupOpen] = React.useState(false);
+  const [isRegisterPopupOpen, setIsRegisterPopupOpen] = React.useState(false);
+  const [isMobileMenuOpen, setIsMobileMenuOpen] = React.useState(false);
+  const [isInfoToolTipOpen, setIsInfoToolTipOpen] = React.useState(false);
+
+  //articles
+
+  const [savedArticles, setSavedArticles] = React.useState([]);
+  const [cards, setCards] = React.useState([]);
+  const [searchData, setSearchData] = React.useState({search: 'Economics'});
+  const [keyword, setKeyword] = React.useState('');
+
+
+  //errors
+
+  const [isServerError, setIsServerError] = React.useState(false);
+  const [isClientError, setIsClientError] = React.useState('');
+
+  //initial api effects
+  //checking token
+
+  React.useEffect(() => {
+    const token = localStorage.getItem('jwt');
+    if (token) {
+      auth
+        .checkTokenValidity(token)
+        .then((res) => {
+          console.log('checking token ', res);
+          if (res._id) { //(res._id)
+            setIsLoggedIn(true);
+            setUserData({ username: res.username });
+            console.log('res.username: ' + res.username + ' current userdata ' + userData);
+          } else {
+            localStorage.removeItem("jwt");
+            setIsLoginPopupOpen(true);
+            setIsLoggedIn(false);
+          }
+        })
+        .catch((err) => {
+          console.log(err);
+          setIsLoginPopupOpen(true);
+          setIsLoggedIn(false);
+          localStorage.removeItem("jwt");
+
+        })
+        .finally(() => {
+          setIsCheckingToken(false);
+        });
+    }
+  }, []);
+
+  //getting user info
+
+  // React.useEffect(() => {
+  //   const token = localStorage.getItem('jwt');
+  //   if (token) {
+  //     mainApi
+  //       .getUserInfo(token)
+  //       .then(res => {
+  //         console.log('getting user info ', res);
+  //         setCurrentUser(res);
+  //       })
+  //       .catch((err) => {
+  //         console.log(err);
+  //       });
+  //   }
+  // }, [isLoggedIn]);
+
+  //getting info on user's saved articles
+
+  React.useEffect(() => {
+    const token = localStorage.getItem('jwt');
+    if (token) {
+      mainApi
+        .getSavedArticles(token) //(token)?
+        .then((res) => {
+          console.log('getting saved articles info ', res);
+          articles.current = res;
+        setSavedArticles(articles.current.slice(0, 3));
+        })
+        .catch((err) => {
+          console.log(err);
+        });
+    }
+  }, [currentUser]);
+
+  
+//handlers  
+//authorization handlers
+
+function handleRegister({ username, email, password }) {
+  auth
```

Comment:
**Might be better**

You can create a special variable for watching **loading** states during server requests:

```jsx
 const [isLoading, setIsLoading] = React.useState(false);

```

And then you can change it to `true` before sending requests and return it back to `false`  in `finally` to stop loading:

And you should pass `isLoading` to every popup to change the **button text**:

```jsx

 buttonText={isLoading? 'Saving...' : 'Save'}

```